proposed/http-handlers/request-handlers-meta.md

```diff
@@ -363,6 +368,237 @@ executing that handler. If unable to route to a handler, it would instead
+    }
+}
+```
+
+An application bootstrap might then look like this:
+
+```php
+// Fallback response:
+$response = (new Response())->withStatus(500);
+
+// Create request handler instance:
+$app = new QueueResponseHandler($response);
+
+// Add one or more middleware:
+$app->add(new CorsMiddleware());
+$app->add(new RoutingMiddleware());
+
+// execute it:
+$response = $app->handle(ServerRequestFactory::fromGlobals());
+```
+
+This system has essentially two request handlers, one that will produce a
+response if no middleware generates one, and one for dispatching the middleware
+layers. (In this example, the `RoutingMiddleware` will likely execute composed
+handlers on a successful route match; see more on that below.)
+
+This approach has the following benefits:
+
+- Middleware does not need to know anything about any other middleware or how it
+  is composed in the application.
+- The `QueueRequestHandler` is agnostic of the PSR-7 implementation in use
+- Middleware is executed in the order it is added to the application, making the
+  code explicit.
+
+#### Decoration-based request handler
+
+In this approach, a request handler implementation decorates both a middleware
+instance and a fallback request handler to pass to it. The application is built
+from the outside-in, passing each request handler "layer" to the next outer one.
+
+```php
+class DecoratedRequestHandler implements RequestHandlerInterface
+{
+    private $middleware;
+    private $nextHandler;
+
+    public function __construct(MiddlewareInterface $middleware, RequestHandlerInterface $nextHandler)
+    {
+        $this->middleware = $middleware;
+        $this->nextHandler = $nextHandler;
+    }
+
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        return $this->middleware->process($request, $this->nextHandler);
+    }
+}
+
+$innerHandler = new class ($responsePrototype) implements RequestHandlerInterface {
+    private $responsePrototype;
+
+    public function __construct(ResponseInterface $responsePrototype)
+    {
+        $this->responsePrototype = $responsePrototype;
+    }
+
+    public function handle(ServerRequestInterface $request) : ResponseInterface
+    {
+        return $this->responsePrototype;
+    }
+};
+
+$layer1 = new DecoratedRequestHandler(new RoutingMiddleware(), $innerHandler);
+$layer2 = new DecoratedRequestHandler(new CorsMiddleware(), $layer1);
+
+$response = $layer2->handle(ServerRequestFactory::fromGlobals());
+```
+
+Similar to the queue-based middleware, request handlers serve two purposes in
+this system:
+
+- Producing a fallback response if no other layer does.
+- Dispatching middleware.
+
+#### Reusable Middleware Examples
+
+In the examples above, we have two middleware composed in each. In order for
+these to work in either situation, we need to write them such that they interact
+appropriately.
+
+Generally speaking, implementors of middleware should follow these guidelines:
```

Comment:
I feel like these guidelines should be moved to the spec proper, in some form, as they directly inform how middlewares will interoperate.  Eg, when to call the next handler cannot be captured in an interface but is still relevant, and if done "wrong" will lead to a non-interoperable implementation.

Reply:
@shadowhand What do you think about moving these into the spec proper? If we do, where should they live?

Reply:
I don't think we should but putting code samples in the spec itself. Having the guidelines at the top of the spec would be good though.

Reply:
Yes, I was referring to the guidelines, (probably written a bit differently).  I agree the code samples should remain in the metadoc.

Reply:
I really think the spec should not dictate *how* middlewares should behave (or other implementation details), e.g. when to generate a response or call the delegate.

Maybe that's not what you meant, but the guidelines below are too precise IMO.

Reply:
I agree with @shadowhand and @mnapoli: I don't think the spec document should contain these guidelines.

Reply:
@mnapoli and @stefanotorresi — thanks; I think that's great feedback, and I tend to agree. The guidelines presented are implementation details and good practices, but _do not need to be followed in order to comply with the specification_. I'll leave them where they are.

Reply:
Could you make that part clear in the text? Because it reads like a specification (with the "should" keyword).

Reply:
I'm not sure what other word to user here, TBH, and the fact that it's not ALL CAPS means it does not have the same semantic weight per RFC 2119. Maybe the following?

> Implementors of middleware striving for maximum interoperability may want to consider the following guidelines:

Would that work for you, @mnapoli ?

Reply:
Updated!